Add support for the HA-PROXY protocol in HTTP(S) requests

vulpine opened the pull request:
I've wanted to be able to curl against proxies which expect the PROXY protocol (http://www.haproxy.org/download/1.5/doc/proxy-protocol.txt), and so have added support for it to curl. This is my first experience of writing C for a project such as this so I expect it probably needs some tidying up, but it works; I can curl against HAProxy which expects PROXY protocol over both HTTP and HTTPS and receive sensible and correct responses.

bagder commented:
Can you explain what a use case could be for supporting this protocol in curl?

Here are some initial nits and questions:

You seem to have based this on an older version of curl, please rebase and fix the conflicts. Also run `make checksrc` and adjust the flaws (which was at least one too long line). I found code commented out, either remove that or use the code! There's no accompanied documentation for the new options. 

`TCP4` is supposedly for IPv4 addresses only, right? But curl is fully IPv6 compliant so are you sure you can do it like that? It seems unlikely to work when IPv6 is used as then `TCP6` should be used according to the spec.

I would also like to have this feature get a test case or three added so that we can make sure it remains functional going forward.

I think we need to do something about the options names. Calling this "the proxy protocol" is mighty confusing since we already have proxy involved in so ways even though I realize that isn't really your fault but the name the haproxy team picked. But all the existing `--proxy-` options about about controlling or instructing a proxy, and there's a bunch of more `--proxy*` options coming with the pending HTTP-proxy work. Maybe call it haproxy protocol?

Finally, please be aware that there's a big chance that this pull request will need to get adjusted again if we manage to pull the HTTPS-proxy into master before we merge this - which is what I intend.

vulpine commented:
@bagder, thank you for the detailed response! Some replies follow:

One use case we have found at Yelp has been where we have our load balancers behind an Amazon Elastic Load Balancer using PROXY protocol to ensure that the remote_ip of users is kept intact. As such, while our load balancers are publicly-routable, they cannot be curled as they will drop the connection as soon as they receive a request which is not prepended with the PROXY protocol line. With this switch, we will be able to curl them directly, which is a great aid in diagnosing issues on particular load balancers.

I will update and rebase against the most recent code, apologies for it being so out of date.

You're right -- TCP4 implies IPv4. I'd not considered IPv6 and will add support. Would it be sensible to check the value of data->conn->bits.ipv6_ip to determine if the request is IPv6, or is there a better way?

I'll work on adding test cases.

With regards to the option name, I was a little concerned about the potential for namespace clash there myself; I'll switch to 'haproxy-protocol', as you suggested, as that should be unambiguous.

vulpine commented:
I think I've addressed more-or-less all your questions from the original review, @bagder. Sorry about the commit mess; I'd be happy to cut a new branch with them all cherry-picked and turned into a single commit if that would make things easier. Otherwise, this now supports both IPv4 and IPv6, has tests, and has been rebased and fixed to not conflict with the state of the art.

bagder commented:
It is even more convenient if you just squash the commits into logical changes yourself (into one or at most a few commits) and push -f them to the branch, then we can review this work easier.

But also: please note that the feature window for next release closed this Monday so we cannot merge this into master until after the pending release. So there's no hurry here, but we can still get it in stellar shape in the mean time and merge as soon as the window opens again.

vulpine commented:
@bagder : No problem, thank you for your help so far - I certainly don't mind waiting. :) I think I have things in better condition now and have squashed it into a single commit.

bagder commented on lib/http.c:
```diff
@@ -1388,6 +1395,37 @@ static int http_getsock_do(struct connectdata *conn,
   return GETSOCK_WRITESOCK(0);
 }
 
+static void add_haproxy_protocol_header(struct connectdata *conn, bool *done)
```
This better return `CURLcode` since it uses functions that can fail and those error are better propagate back up.

bagder commented on lib/http.c:
```diff
@@ -1388,6 +1395,37 @@ static int http_getsock_do(struct connectdata *conn,
+{
+  char proxy_header[128];
+  Curl_send_buffer *req_buffer;
+  CURLcode result;
```
This compiler warning is of course because your function is 'void' and you never do anything with the return values in the function...
~~~
http.c:1404:12: error: variable ‘result’ set but not used [-Werror=unused-but-set-variable]
   CURLcode result;

bagder commented on lib/http.c:
```diff
@@ -1388,6 +1395,37 @@ static int http_getsock_do(struct connectdata *conn,
+  /* Emit the correct prefix for IPv6 */
+  if (conn->bits.ipv6) {
+    strcpy(tcp_version, "TCP6");
+  } else {
```
We don't write 'else' clauses like that: (both these style nits were found with `make checksrc` )
~~~
./http.c:1410:3: warning: else after closing brace on same line (BRACEELSE)
   } else {

bagder commented:
ping @vulpine ?

vulpine commented:
Apologies -- have been super busy. Thank you for the comments, I will try to get to them this week!

> On 11 Feb 2017, at 16:59, Daniel Stenberg <notifications@github.com> wrote:
> 
> ping @vulpine ?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

vulpine commented:
@bagder - I've made changes per your comments. `make checksrc` is now clean. Apologies this took so long.

noahwilliamsson commented:
Hey @vulpine, nice work implementing support for the HAProxy PROXY protocol.  I've got a similar setup as you and face the same issues debugging https services that require a PROXY protocol header.  I would be happy to see this merged.

Any chance you have time for a final cleanup?  Is there anything I can do to help?

bagder commented:
We obviously *at least* need the work rebased on top of the current git master without conflicts so that we can again review it and properly consider it for merging.

elyscape commented:
@bagder I went ahead and opened #2162, in which I rebased this pull request onto master, fixed code sections that have changed, and added some additional error checking that was missing from this code.

bagder commented:
#2162 has the current work on this
